tf_quant_finance/models/heston/approximations/asian_prices.py

```diff
@@ -0,0 +1,509 @@
+      for the payoff calculation
+      Default value: AveragingType.GEOMETRIC
+    averaging_frequency: Enum value of AveragingFrequency to select the
+      averaging type for the payoff calculation (discrete vs continuous)
+      Default value: AveragingFrequency.DISCRETE
+    integration_method: An instance of `math.integration.IntegrationMethod`.
+      Default value: `None` which maps to Gaussian quadrature.
+    dtype: Optional `tf.DType`. If supplied, the dtype to be used for conversion
+      of any supplied non-`Tensor` arguments to `Tensor`.
+      Default value: None which maps to the default dtype inferred by
+      TensorFlow.
+    name: str. The name for the ops created by this function.
+      Default value: `None` which is mapped to the default name 
+      `discrete_geometric_average_price`.
+    **kwargs: Additional parameters for the underlying integration method.
+      If not supplied, uses bounds `lower=1e-9`, `upper=100`.
+
+  Returns:
+    option_prices: A `Tensor` of the same shape as `strikes`. The Heston price
+    of the options.
+
+  Raises:
+    ValueError: If both `forwards` and `spots` are supplied or if neither is
+      supplied.
+    ValueError: If both `discount_rates` and `discount_factors` is supplied.
+    ValueError: If option is arithmetic.
+    NotImplementedError: if option is continuous averaging.
+    NotImplementedError: if any of the Heston model parameters are of type
+      PiecewiseConstantFunc.
+  """
+  if (spots is None) == (forwards is None):
+    raise ValueError('Either spots or forwards must be supplied but not both.')
+  if (discount_rates is not None) and (discount_factors is not None):
+    raise ValueError('At most one of discount_rates and discount_factors may '
+                     'be supplied')
+  if averaging_type == AveragingType.ARITHMETIC:
+    raise ValueError('Cannot price arithmetic averaging asians analytically '
+                     'under Heston model')
+  if averaging_frequency == AveragingFrequency.DISCRETE:
+    if sampling_times is None:
+      raise ValueError('Sampling times required for discrete sampling asians')
+  if averaging_frequency == AveragingFrequency.CONTINUOUS:
+    raise NotImplementedError('Pricing continuous averaging asians not yet '
+                              'supported')
+
+  with tf.name_scope(name or 'asian_option_price'):
+    strikes = tf.convert_to_tensor(strikes, dtype=dtype, name='strikes')
+    dtype = strikes.dtype
+    results_shape = strikes.shape
+
+    expiries = tf.convert_to_tensor(expiries, dtype=dtype, name='expiries')
+    expiries = tf.broadcast_to(expiries, strikes.shape)
+    sampling_times = tf.convert_to_tensor(
+        sampling_times, dtype=dtype, name='sampling_times')
+    sampling_times = tf.broadcast_to(
+        sampling_times, [sampling_times.shape[0], strikes.shape[0]])
+
+    # Fail assertion if sampling comes after expiry
+    if averaging_frequency == AveragingFrequency.DISCRETE:
```

Comment:
maybe raise NotImplementedError when  `averaging_frequency` is not discrete?

Reply:
There in lines 189-190 of the latest version